// webpack-digest/lib/config.js
/*

Copyright (C) 2016,2017 wallzero 2016,2017

This program is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software Foundation,
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with this
program. If not, see <http://www.gnu.org/licenses/>.

*/

/*
  eslint-disable
  import/unambiguous,
  import/no-commonjs,
  no-process-env,
  no-nested-ternary,
  import/no-dynamic-require
*/

const path = require('path');

const resolve = require(path.join(__dirname, 'resolve.js'));
const production = process.env.NODE_ENV !== 'development';

const digestPath = resolve('webpack.digest.config.json', process.env.DIGEST) ||
  resolve('webpack.digest.config.js', process.env.DIGEST);
const digest = digestPath ? Object.assign({platforms: {}}, require(digestPath)) : {platforms: {}}; // eslint-disable-line global-require

const srcPath = process.env.SRC_PATH || digest.srcPath || 'src';
const library = process.env.LIBRARY === 'true' || process.env.LIBRARY === true || digest.library === true;
const platformsArray = process.env.PLATFORMS ? process.env.PLATFORMS.split(',') : [];
const project = require(resolve('package.json')); // eslint-disable-line global-require

module.exports = {
  baseHref: process.env.BASE_HREF || digest.baseHref || '/',
  cache: path.join(
    process.env.CACHE || digest.cache || path.join(process.cwd(), 'node_modules', '.cache'),
    library ? 'library' : production ? 'production' : 'development'
  ),
  compression: process.env.COMPRESSION !== 'off' && digest.compression !== 'off',
  config: resolve('webpack.config.js', process.env.CONFIG || digest.config),
  context: path.join(process.cwd(), srcPath || digest.srcPath),
  dashboard: process.env.DASHBOARD !== 'off' && digest.dashboard !== 'off',
  e2e: process.env.E2E_PATH || digest.e2ePath || '__e2e__', // eslint-disable-line id-match
  env: {
    // eslint-disable-next-line no-extra-parens
    ...(process.env.ENV && JSON.parse(process.env.ENV.replace(/""(.*)""/g, '"\\"$1\\""')) || {}),
    ...digest.env
  },
  eslint: resolve('.eslintrc', process.env.ESLINT || digest.eslint),
  favicon: resolve('favicon.png', process.env.FAVICON || digest.favicon),
  headless: !(
    process.env.HEADLESS === 'false' ||
    process.env.HEADLESS === false ||
    digest.headless === false ||
    false
  ),
  ignore: new RegExp('(node_modules|lib)[\\/\\\\]'),
  index: resolve('index.htm', process.env.INDEX || digest.index),
  library,
  parallel: process.env.PARALLEL !== 'off' && digest.parallel !== 'off',
  platforms: {
    angular: platformsArray.includes('angular') || digest.platforms.angular,
    cssModules: platformsArray.includes('cssModules') || digest.platforms.cssModules,
    flow: platformsArray.includes('flow') || digest.platforms.flow,
    pragma: platformsArray.includes('react') ? 'react' : digest.platforms.pragma,
    typescript: platformsArray.includes('typescript') || digest.platforms.typescript
  },
  polyfills: resolve('polyfills.js', process.env.POLYFILLS || digest.polyfills),
  port: process.env.PORT ? process.env.PORT : production ? 8080 : 8090,
  postcss: resolve('postcss.config.js', process.env.POSTCSS || digest.postcss),
  postcssWhitelist: process.env.POSTCSS_WHITELIST || digest.postcssWhitelist || 'react-toolbox',
  production,
  project,
  pwa: process.env.PWA !== 'off' && digest.pwa !== 'off',
  reportPath: process.env.REPORT_PATH || digest.reportPath ?
    path.join(process.cwd(), process.env.REPORT_PATH || digest.reportPath) :
    path.join(process.cwd(), 'reports'),
  resolve,
  seleniumHost: process.env.SELENIUM_HOST || digest.seleniumHost || '127.0.0.1',
  seleniumPort: process.env.SELENIUM_PORT || digest.seleniumPort || '4444',
  seleniumStart: !(
    process.env.SELENIUM_START === 'false' ||
    process.env.SELENIUM_START === false ||
    digest.seleniumStart === false ||
    false
  ),
  shortName: process.env.SHORT_NAME || digest.shortName || project.name || 'wd',
  sourceMap: process.env.SOURCE_MAP !== 'false' && process.env.SOURCE_MAP !== false && digest.sourceMap !== false,
  srcPath,
  staticPath: process.env.STATIC_PATH || digest.staticPath ?
    path.join(process.cwd(), process.env.STATIC_PATH || digest.staticPath, library ? '' : 'bundle') :
    path.join(process.cwd(), library ? 'lib' : 'public', library ? '' : 'bundle'),
  stylelint: resolve('.stylelintrc', process.env.STYLELINT || digest.stylelint),
  stylelintIgnore: resolve('.stylelintrc', process.env.STYLELINT_IGNORE || digest.stylelintIgnore),
  testConfig: resolve('testconfig.js', process.env.TESTCONFIG || digest.testConfig),
  tsconfig: resolve('tsconfig.json', process.env.TSCONFIG || digest.tsconfig),
  tslint: resolve('tslint.json', process.env.TSLINT || digest.tslint)
};

/*
  eslint-enable
  import/unambiguous,
  import/no-commonjs,
  no-process-env,
  no-nested-ternary,
  import/no-dynamic-require
*/
