// webpack-digest/lib/jest/config.js
/*

Copyright (C) 2016,2017 wallzero 2016,2017

This program is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software Foundation,
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with this
program. If not, see <http://www.gnu.org/licenses/>.

*/

/*
  eslint-disable
  import/unambiguous,
  import/no-commonjs,
  no-useless-escape,
  max-len
*/

const path = require('path');
const {
  reportPath,
  testConfig
} = require('../config');

module.exports = {
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
    'demo/**/*.{js,jsx,ts,tsx}',
    '!**/__e2e__/**/*',
    '!**/__tests__/**/*',
    '!src/**/*.{d.ts}',
    '!demo/**/*.{d.ts}'
  ],
  coverageDirectory: path.join(reportPath, 'coverage'),
  moduleDirectories: ['node_modules', '<rootDir>', 'node_modules/webpack-digest/node_modules'],
  moduleFileExtensions: ['json', 'js', 'jsx', 'ts', 'tsx'],
  moduleNameMapper: {
    '\\.(css|scss|sass|less)$': 'identity-obj-proxy'
  },
  rootDir: '../../../../',
  setupFiles: [
    '<rootDir>/node_modules/webpack-digest/lib/jest/setup.js',
    testConfig
  ],
  snapshotSerializers: [
    '<rootDir>/node_modules/enzyme-to-json/serializer'
  ],
  testPathIgnorePatterns: ['/node_modules/', '/lib/', '/public/', '/static/', '/bin/', '.d.ts'],
  testRegex: '(__tests__[\\\/\\\\].*|\\.(test|spec))\\.(j|t)sx?$',
  transform: {
    '\\.(jpg|jpeg|png|gif|eot|otf|webp|svg|ttf|woff|woff2|mp4|webm|wav|mp3|m4a|aac|oga)(\\?v=[0-9]\\.[0-9]\\.[0-9])?$': '<rootDir>/node_modules/webpack-digest/lib/jest/transformMisc.js',
    '^.+\\.js$': '<rootDir>/node_modules/webpack-digest/lib/jest/transformJs.js',
    '^.+\\.jsx$': '<rootDir>/node_modules/webpack-digest/lib/jest/transformJsx.js',
    '^.+\\.ts$': '<rootDir>/node_modules/webpack-digest/lib/jest/transformTs.js',
    '^.+\\.tsx$': '<rootDir>/node_modules/webpack-digest/lib/jest/transformTsx.js'
  }
};

/*
  eslint-enable
  import/unambiguous,
  import/no-commonjs,
  no-useless-escape,
  max-len
*/
