// webpack-digest/lib/nightwatch/vars.conf.js
/*

Copyright (C) 2016,2017 wallzero 2016,2017

This program is free software: you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software Foundation,
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.
You should have received a copy of the GNU General Public License along with this
program. If not, see <http://www.gnu.org/licenses/>.

*/

/*
  eslint-disable
  global-require,
  no-process-env,
  no-undefined,
  no-nested-ternary,
  complexity,
  import/no-commonjs,
  import/unambiguous,
  import/no-dynamic-require,
  import/max-dependencies,
  filenames/match-regex
*/

const path = require('path');
const {
  reportPath
} = require('../config'); // eslint-disable-line id-match

const binPath = path.join(process.cwd(), 'node_modules', 'nightwatch', 'bin');

// https://selenium-release.storage.googleapis.com/index.html
const seleniumVersion = '3.10.0';

// https://chromedriver.storage.googleapis.com/index.html
const chromeDriverVersion = '2.37';

// https://github.com/mozilla/geckodriver/releases
const geckoDriverVersion = '0.20.1';

module.exports = {
  binPath,

  // https://selenium-release.storage.googleapis.com/index.html
  chromeDriverVersion,
  chromePath: path.join(binPath, 'chromedriver', chromeDriverVersion + '-' + process.arch + '-chromedriver'),

  // https://github.com/mozilla/geckodriver/releases
  geckoDriverVersion,
  geckoPath: path.join(binPath, 'geckodriver', geckoDriverVersion + '-' + process.arch + '-geckodriver'),

  // https://selenium-release.storage.googleapis.com/index.html
  ieDriverVersion: seleniumVersion,

  reportsPath: path.join(reportPath, 'e2e'),
  screenshotPath: path.join(reportPath, 'screenshots'),

  // https://selenium-release.storage.googleapis.com/index.html
  seleniumHost: '127.0.0.1',
  seleniumPath: path.join(binPath, 'selenium-server', seleniumVersion + '-server.jar'),
  seleniumVersion
};

/*
  eslint-enable
  global-require,
  no-process-env,
  no-undefined,
  no-nested-ternary,
  complexity,
  import/no-commonjs,
  import/unambiguous,
  import/no-dynamic-require,
  import/max-dependencies,
  filenames/match-regex
*/
